helpers/selfproxy.py

```diff
@@ -0,0 +1,201 @@
+        self.out_buffer = self.out_buffer[num_sent:]
+        self.last_segment_time = time.time()
+        if len(self.out_buffer) == 0 and self.closing:
+            self.handle_close()
+
+    def in_pause(self):
+        return ( self.do_chunking and self.proxy.segment_gap > 0 and
+                 time.time() - self.last_segment_time
+                     < self.proxy.segment_gap / 1000.0 )
+
+    def writable(self):
+        if self.in_pause():
+            return False;
+        if self.closing:
+            return True;
+        return asyncore.dispatcher_with_send.writable(self)
+
+    def send(self, data):
+        self.out_buffer = self.out_buffer + data
+        if not self.in_pause():
+            self.initiate_send()
+
+    def readable(self):
+        return self.pair.connected
+
+    def handle_read(self):
+        self.pair.send(self.recv(MAX_MESSAGE_SIZE))
+
+    def handle_close(self):
+        self.closing = True
+        self.pair.closing = True
+        self.close()
+
+class SelfProxy(asyncore.dispatcher):
+
+    def __init__(self, mode, listen_host, listen_port,
+                       forward_host, forward_port,
+                       segment_size, segment_gap):
+        asyncore.dispatcher.__init__(self)
+        self.mode = mode
+        self.listen_host = listen_host
+        self.listen_port = listen_port
+        self.forward_host = forward_host
+        self.forward_port = forward_port
+        self.segment_size = segment_size
+        self.segment_gap = segment_gap
+        self.connections = []
+
+        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.set_reuse_addr()
+        self.bind((self.listen_host, self.listen_port))
+        self.listen(socket.SOMAXCONN)
+
+    def handle_close(self):
+        self.close()
+        for conn in self.connections:
+            if conn.connected:
+                conn.handle_close()
+
+    def stop(self):
+        self.handle_close()
+
+    def handle_accept(self):
+        pair = self.accept()
+        if pair is not None:
+            sock, _ = pair
+            accepted_conn = ProxyConnection(sock=sock, proxy=self)
+            forward_conn = ProxyConnection(pair=accepted_conn, proxy=self)
+            accepted_conn.pair = forward_conn
+            if self.mode == CLIENT_MODE:
+                forward_conn.do_chunking = True
+            elif self.mode == SERVER_MODE:
+                accepted_conn.do_chunking = True
+                if self.segment_size:
+                    accepted_conn.socket.setsockopt(socket.SOL_TCP,
+                        socket.TCP_NODELAY, 1)
+            else:
+                error.bug("Invalid SelfProxy mode")
+            self.connections.append(accepted_conn)
+            self.connections.append(forward_conn)
+            forward_conn.create_socket(socket.AF_INET, socket.SOCK_STREAM)
+            forward_conn.bind((self.listen_host, 0))
+            forward_conn.connect((self.forward_host, self.forward_port))
```

Comment:
Thank you very much!
This is strange. I'll think on it.

> Also, for some reason the problematic code is triggered by very few tests

Because it is too new tool